uniffi_macros/src/export.rs

```diff
@@ -42,12 +44,48 @@ pub(crate) fn expand_export(
                         gen_constructor_scaffolding(sig, &mod_path, &self_ident, &args)
                     }
                     ImplItem::Method(sig) => {
-                        gen_method_scaffolding(sig, &mod_path, &self_ident, &args)
+                        gen_method_scaffolding(sig, &mod_path, &self_ident, &args, false)
                     }
                 })
                 .collect::<syn::Result<_>>()?;
             Ok(quote_spanned! { self_ident.span() => #item_tokens })
         }
+        ExportItem::Trait { items, self_ident } => {
+            let name = ident_to_string(&self_ident);
+            let free_fn_ident =
+                Ident::new(&free_fn_symbol_name(&mod_path, &name), Span::call_site());
+
+            let free_tokens: syn::Result<TokenStream> = Ok(quote! {
+                #[doc(hidden)]
+                #[no_mangle]
+                pub extern "C" fn #free_fn_ident(
+                    ptr: *const ::std::ffi::c_void,
+                    call_status: &mut ::uniffi::RustCallStatus
+                ) {
+                    uniffi::rust_call(call_status, || {
+                        assert!(!ptr.is_null());
+                        drop(unsafe { Box::from_raw(ptr as *mut std::sync::Arc<dyn #self_ident>) });
+                        Ok(())
+                    });
+                }
+            });
+
+            let impl_tokens = items.into_iter().map(|item| match item? {
+                ImplItem::Method(sig) => {
+                    gen_method_scaffolding(sig, &mod_path, &self_ident, &args, true)
+                }
+                _ => unreachable!("traits have no constructors"),
+            });
+            let meta_static_var = interface_meta_static_var(&self_ident, true, &mod_path);
+            let macro_tokens = Ok(quote! {
+                ::uniffi::ffi_converter_trait_decl!(dyn #self_ident, stringify!(#self_ident), crate::UniFfiTag);
+            });
+            let item_tokens: TokenStream = vec![meta_static_var, free_tokens, macro_tokens]
+                .into_iter()
+                .chain(impl_tokens)
+                .collect::<syn::Result<_>>()?;
```

Comment:
What's the idea behind doing all the work, then returning the first error found?

Reply:
TBH I initially just copy-pasted that from a few lines above, then rolled with it. I *think* I've made the change you suggest.